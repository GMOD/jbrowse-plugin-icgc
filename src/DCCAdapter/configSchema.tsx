import { ConfigurationSchema } from '@jbrowse/core/configuration'

export default ConfigurationSchema(
  'DCCAdapter',
  {
    dccLocation: {
      type: 'fileLocation',
      defaultValue: { uri: '/path/file.tsv', locationType: 'UriLocation' },
    },
  },
  { explicitlyTyped: true },
)
