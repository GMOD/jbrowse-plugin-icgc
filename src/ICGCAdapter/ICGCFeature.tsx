import {
  SimpleFeatureSerialized,
  Feature,
} from '@jbrowse/core/util/simpleFeature'

interface FeatureData {
  [key: string]: unknown
  refName: string
  start: number
  end: number
  type: string
}

export default class ICGCFeature implements Feature {
  private icgcObject: any

  private data: FeatureData

  private uniqueId: string

  private featureType: string

  constructor(args: {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    icgcObject: any
    id: string
    featureType: string
  }) {
    this.icgcObject = args.icgcObject
    this.featureType = args.featureType ? args.featureType : 'mutation'
    this.data = this.dataFromICGCObject(this.icgcObject, this.featureType)
    this.uniqueId = args.id
  }

  get(field: string) {
    return this.icgcObject[field] || this.data[field]
  }

  set() {}

  parent() {
    return undefined
  }

  children() {
    return undefined
  }

  tags(): string[] {
    const t = [...Object.keys(this.data), ...Object.keys(this.icgcObject)]
    return t
  }

  id(): string {
    return this.uniqueId
  }

  dataFromICGCObject(icgcObject: any, featureType: string): FeatureData {
    // Defaults to mutation values
    const featureData: FeatureData = {
      refName: icgcObject.chromosome,
      type: icgcObject.mutationType,
      start: icgcObject.start - 1,
      end: icgcObject.end,
    }

    if (featureType === 'occurrences') {
      featureData.type = icgcObject.mutationId
      featureData.id = icgcObject.donorId
      featureData.note = icgcObject.mutationId
    }

    return featureData
  }

  toJSON(): SimpleFeatureSerialized {
    return {
      uniqueId: this.uniqueId,
      ...this.data,
      ...this.icgcObject,
    }
  }
}
