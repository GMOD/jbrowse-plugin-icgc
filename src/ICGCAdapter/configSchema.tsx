import { ConfigurationSchema } from '@jbrowse/core/configuration'
import { types } from 'mobx-state-tree'

export default ConfigurationSchema(
  'ICGCAdapter',
  {},
  { explicitlyTyped: true, explicitIdentifier: 'ICGCAdapterId' },
)
