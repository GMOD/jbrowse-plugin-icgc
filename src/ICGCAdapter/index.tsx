export { default as AdapterClass } from './ICGCAdapter'
export { default as configSchema } from './configSchema'
