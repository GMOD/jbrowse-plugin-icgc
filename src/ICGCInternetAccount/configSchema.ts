import { ConfigurationSchema } from '@jbrowse/core/configuration'
import { Instance } from 'mobx-state-tree'
import { BaseInternetAccountConfig } from '@jbrowse/core/pluggableElementTypes/models'

const ICGCConfigSchema = ConfigurationSchema(
  'ICGCInternetAccount',
  {},
  {
    baseConfiguration: BaseInternetAccountConfig,
    explicitlyTyped: true,
  },
)

export type ICGCInternetAccountConfigModel = typeof ICGCConfigSchema
export type OAuthInternetAccountConfig = Instance<
  ICGCInternetAccountConfigModel
>
export default ICGCConfigSchema
