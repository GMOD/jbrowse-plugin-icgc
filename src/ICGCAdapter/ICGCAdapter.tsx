import { BaseFeatureDataAdapter } from '@jbrowse/core/data_adapters/BaseAdapter'
import { ObservableCreate } from '@jbrowse/core/util/rxjs'
import { Region } from '@jbrowse/core/util/types'
import { Feature } from '@jbrowse/core/util/simpleFeature'
import { Instance } from 'mobx-state-tree'
import { readConfObject } from '@jbrowse/core/configuration'
import { BaseOptions } from '@jbrowse/core/data_adapters/BaseAdapter'
import ICGCFeature from './ICGCFeature'
import MyConfigSchema from './configSchema'
import AbortablePromiseCache from 'abortable-promise-cache'
import LRU from '@jbrowse/core/util/QuickLRU'

export default class ICGCAdapter extends BaseFeatureDataAdapter {
  private filters: string

  private size: number

  private featureType: string

  public static capabilities = ['getFeatures', 'getRefNames']

  public constructor(config: Instance<typeof MyConfigSchema>) {
    super(config)

    const filters = readConfObject(config, 'filters') as string
    const size = readConfObject(config, 'size') as number
    const featureType = readConfObject(config, 'featureType') as string

    this.filters = filters
    this.size = size
    this.featureType = featureType
  }

  private featureCache = new AbortablePromiseCache({
    cache: new LRU({ maxSize: 200 }),
    fill: async (query: any, abortSignal?: AbortSignal) => {
      return this.fetchFeatures(query, abortSignal)
    },
  })

  private async fetchFeatures(query: string, signal?: AbortSignal) {
    const response = await fetch(
      `http://localhost:7080/proxy/api/v1/${this.featureType}?filters=${query}&size=${this.size}`,
      {
        method: 'GET',
        headers: { 'content-type': 'application/json' },
        signal,
      },
    )
    if (!response.ok) {
      throw new Error(
        `Failed to fetch ${response.status} ${response.statusText}`,
      )
    }
    return response.json()
  }

  public async getRefNames() {
    return [
      'chr1',
      'chr10',
      'chr11',
      'chr12',
      'chr13',
      'chr14',
      'chr15',
      'chr16',
      'chr17',
      'chr18',
      'chr19',
      'chr2',
      'chr20',
      'chr21',
      'chr22',
      'chr3',
      'chr4',
      'chr5',
      'chr6',
      'chr7',
      'chr8',
      'chr9',
      'chrX',
      'chrY',
    ]
  }

  private createQuery(location: string, start: number, end: number) {
    let query = {}
    if (this.filters != '{}') {
      const mutation = JSON.parse(this.filters)['mutation']
      const gene = JSON.parse(this.filters)['gene']
      const donor = JSON.parse(this.filters)['donor']

      if (mutation) {
        query = {
          mutation: {
            ...mutation,
            location: {
              is: `${location}:${start}-${end}`,
            },
          },
        }
      } else {
        query = {
          mutation: {
            location: {
              is: `${location}:${start}-${end}`,
            },
          },
        }
      }
      if (gene) {
        query = {
          ...query,
          gene,
        }
      }
      if (donor) {
        query = {
          ...query,
          donor,
        }
      }
    } else {
      query = {
        mutation: {
          location: {
            is: `${location}:${start}-${end}`,
          },
        },
      }
    }
    return query
  }

  public getFeatures(region: Region, opts: BaseOptions = {}) {
    const { refName, start, end } = region

    return ObservableCreate<Feature>(async (observer) => {
      try {
        const query = this.createQuery(refName, start, end)
        // idField for occurrences is donorId
        const idField = this.featureType === 'mutations' ? 'id' : 'donorId'

        const result = await this.featureCache.get(
          JSON.stringify(query),
          JSON.stringify(query),
          opts.signal,
        )

        for (const hit of result.hits) {
          const feature = new ICGCFeature({
            icgcObject: hit,
            id: hit[idField] as string,
            featureType: this.featureType,
          })
          observer.next(feature)
        }
      } catch (e) {
        observer.error(e)
      }
      observer.complete()
    }, opts.signal)
  }

  public freeResources(): void {}
}
