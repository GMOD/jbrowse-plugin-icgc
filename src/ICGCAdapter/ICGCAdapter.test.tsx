import { toArray } from 'rxjs/operators'
import ICGCAdapter from './ICGCAdapter'
import configSchema from './configSchema'
import { icgcFilters, icgcResponse } from './test_data/icgc_test_data.jsx'
import fetchMock from 'jest-fetch-mock'

fetchMock.enableMocks()

test('adapter can fetch features from the icgc', async () => {
  const adapter = new ICGCAdapter(
    configSchema.create({
      ICGCAdapterId: 'test-icgc-adapter',
      filters: JSON.stringify(icgcFilters),
    }),
  )

  fetchMock.mockResponseOnce(JSON.stringify(icgcResponse))

  const features = adapter.getFeatures({
    assemblyName: 'volvox',
    refName: 'chr3',
    start: 124900600,
    end: 124925000,
  })

  const names = await adapter.getRefNames()
  expect(names).toMatchSnapshot()

  const featuresArray = await features.pipe(toArray()).toPromise()
  expect(featuresArray.slice(0, 3)).toMatchSnapshot()
})
