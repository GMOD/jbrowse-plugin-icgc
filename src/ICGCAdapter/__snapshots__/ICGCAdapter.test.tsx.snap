// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`adapter can fetch features from the icgc 1`] = `
Array [
  "chr1",
  "chr10",
  "chr11",
  "chr12",
  "chr13",
  "chr14",
  "chr15",
  "chr16",
  "chr17",
  "chr18",
  "chr19",
  "chr2",
  "chr20",
  "chr21",
  "chr22",
  "chr3",
  "chr4",
  "chr5",
  "chr6",
  "chr7",
  "chr8",
  "chr9",
  "chrX",
  "chrY",
]
`;

exports[`adapter can fetch features from the icgc 2`] = `
Array [
  Object {
    "affectedDonorCountFiltered": 2,
    "affectedDonorCountTotal": 6,
    "affectedProjectCount": 1,
    "assemblyVersion": "GRCh37",
    "chromosome": "3",
    "end": 124921497,
    "functionalImpact": Array [
      "Unknown",
      "Unknown",
      "Unknown",
      "Unknown",
      "Unknown",
    ],
    "id": "MU89172348",
    "mutation": "->AAAAAAA",
    "refName": "3",
    "referenceGenomeAllele": "-",
    "start": 124921497,
    "study": Array [],
    "testedDonorCount": 67,
    "transcripts": Array [
      Object {
        "consequence": Object {
          "functionalImpact": "Unknown",
          "type": "intron_variant",
        },
        "functionalImpact": "Unknown",
        "id": "ENST00000314584",
      },
      Object {
        "consequence": Object {
          "functionalImpact": "Unknown",
          "type": "intron_variant",
        },
        "functionalImpact": "Unknown",
        "id": "ENST00000393469",
      },
      Object {
        "consequence": Object {
          "functionalImpact": "Unknown",
          "type": "intron_variant",
        },
        "functionalImpact": "Unknown",
        "id": "ENST00000423114",
      },
      Object {
        "consequence": Object {
          "functionalImpact": "Unknown",
          "type": "intron_variant",
        },
        "functionalImpact": "Unknown",
        "id": "ENST00000469902",
      },
      Object {
        "consequence": Object {
          "functionalImpact": "Unknown",
          "type": "downstream_gene_variant",
        },
        "functionalImpact": "Unknown",
        "id": "ENST00000481760",
      },
    ],
    "type": "insertion of <=200bp",
    "uniqueId": "MU89172348",
  },
  Object {
    "affectedDonorCountFiltered": 2,
    "affectedDonorCountTotal": 2,
    "affectedProjectCount": 2,
    "assemblyVersion": "GRCh37",
    "chromosome": "3",
    "end": 124918162,
    "functionalImpact": Array [
      "Unknown",
      "Unknown",
      "Unknown",
      "Unknown",
    ],
    "id": "MU30653737",
    "mutation": "C>T",
    "refName": "3",
    "referenceGenomeAllele": "C",
    "start": 124918162,
    "study": Array [
      "PCAWG",
    ],
    "testedDonorCount": 137,
    "transcripts": Array [
      Object {
        "consequence": Object {
          "functionalImpact": "Unknown",
          "type": "intron_variant",
        },
        "functionalImpact": "Unknown",
        "id": "ENST00000314584",
      },
      Object {
        "consequence": Object {
          "functionalImpact": "Unknown",
          "type": "intron_variant",
        },
        "functionalImpact": "Unknown",
        "id": "ENST00000393469",
      },
      Object {
        "consequence": Object {
          "functionalImpact": "Unknown",
          "type": "intron_variant",
        },
        "functionalImpact": "Unknown",
        "id": "ENST00000423114",
      },
      Object {
        "consequence": Object {
          "functionalImpact": "Unknown",
          "type": "intron_variant",
        },
        "functionalImpact": "Unknown",
        "id": "ENST00000469902",
      },
    ],
    "type": "single base substitution",
    "uniqueId": "MU30653737",
  },
  Object {
    "affectedDonorCountFiltered": 1,
    "affectedDonorCountTotal": 1,
    "affectedProjectCount": 1,
    "assemblyVersion": "GRCh37",
    "chromosome": "3",
    "end": 124897024,
    "functionalImpact": Array [
      "Unknown",
      "Unknown",
      "Unknown",
      "Unknown",
      "Unknown",
      "Unknown",
      "Unknown",
    ],
    "id": "MU30416030",
    "mutation": "A>T",
    "refName": "3",
    "referenceGenomeAllele": "A",
    "start": 124897024,
    "study": Array [],
    "testedDonorCount": 140,
    "transcripts": Array [
      Object {
        "consequence": Object {
          "functionalImpact": "Unknown",
          "type": "intron_variant",
        },
        "functionalImpact": "Unknown",
        "id": "ENST00000314584",
      },
      Object {
        "consequence": Object {
          "functionalImpact": "Unknown",
          "type": "intron_variant",
        },
        "functionalImpact": "Unknown",
        "id": "ENST00000393469",
      },
      Object {
        "consequence": Object {
          "functionalImpact": "Unknown",
          "type": "intron_variant",
        },
        "functionalImpact": "Unknown",
        "id": "ENST00000423114",
      },
      Object {
        "consequence": Object {
          "functionalImpact": "Unknown",
          "type": "intron_variant",
        },
        "functionalImpact": "Unknown",
        "id": "ENST00000462437",
      },
      Object {
        "consequence": Object {
          "functionalImpact": "Unknown",
          "type": "intron_variant",
        },
        "functionalImpact": "Unknown",
        "id": "ENST00000469902",
      },
      Object {
        "consequence": Object {
          "functionalImpact": "Unknown",
          "type": "intron_variant",
        },
        "functionalImpact": "Unknown",
        "id": "ENST00000473262",
      },
      Object {
        "consequence": Object {
          "functionalImpact": "Unknown",
          "type": "intron_variant",
        },
        "functionalImpact": "Unknown",
        "id": "ENST00000479826",
      },
    ],
    "type": "single base substitution",
    "uniqueId": "MU30416030",
  },
]
`;
